src/sources/host_metrics/mod.rs

```diff
@@ -210,144 +210,129 @@ impl HostMetrics {
-                    timestamp,
-                    time.get::<second>() as f64,
-                    BTreeMap::default(),
-                ));
-            }
+            Ok(time) => output.gauge("uptime", time.get::<second>() as f64, BTreeMap::default()),
             Err(error) => {
                 error!(message = "Failed to load host uptime info.", %error, internal_log_rate_secs = 60);
             }
         }
 
         match heim::host::boot_time().await {
-            Ok(time) => {
-                let timestamp = Utc::now();
-                metrics.push(self.gauge(
-                    "boot_time",
-                    timestamp,
-                    time.get::<second>() as f64,
-                    BTreeMap::default(),
-                ));
-            }
+            Ok(time) => output.gauge(
+                "boot_time",
+                time.get::<second>() as f64,
+                BTreeMap::default(),
+            ),
             Err(error) => {
                 error!(message = "Failed to load host boot time info.", %error, internal_log_rate_secs = 60);
             }
         }
+    }
+}
 
-        metrics
+#[derive(Default)]
+pub struct MetricsBuffer {
+    pub metrics: Vec<Metric>,
+    name: &'static str,
+    host: Option<String>,
+    timestamp: DateTime<Utc>,
+    namespace: Option<String>,
+}
+
+impl MetricsBuffer {
+    fn new(namespace: Option<String>) -> Self {
+        Self {
+            metrics: Vec::new(),
+            name: "",
+            host: crate::get_hostname().ok(),
+            timestamp: Utc::now(),
+            namespace,
+        }
+    }
+
+    fn tags(&self, mut tags: MetricTags) -> MetricTags {
+        tags.insert("collector".into(), self.name.into());
+        if let Some(host) = &self.host {
+            tags.insert("host".into(), host.clone());
+        }
+        tags
     }
 
-    fn counter(
-        &self,
-        name: &str,
-        timestamp: DateTime<Utc>,
-        value: f64,
-        tags: BTreeMap<String, String>,
-    ) -> Metric {
-        Metric::new(name, MetricKind::Absolute, MetricValue::Counter { value })
-            .with_namespace(self.config.namespace.clone())
-            .with_tags(Some(tags))
-            .with_timestamp(Some(timestamp))
-    }
-
-    fn gauge(
-        &self,
-        name: &str,
-        timestamp: DateTime<Utc>,
-        value: f64,
-        tags: BTreeMap<String, String>,
-    ) -> Metric {
-        Metric::new(name, MetricKind::Absolute, MetricValue::Gauge { value })
-            .with_namespace(self.config.namespace.clone())
-            .with_tags(Some(tags))
-            .with_timestamp(Some(timestamp))
+    fn counter(&mut self, name: &str, value: f64, tags: MetricTags) {
+        self.metrics.push(
+            Metric::new(name, MetricKind::Absolute, MetricValue::Counter { value })
+                .with_namespace(self.namespace.clone())
+                .with_tags(Some(self.tags(tags)))
+                .with_timestamp(Some(self.timestamp)),
```

Comment:
I would expect this to just be `now()` to keep the timestamp as close as possible to the collection time. Am I missing something though?

Reply:
Each call to `Utc::now()` calls the system call `gettimeofday`, which would then happen once for each metric. While that is optimized on Linux, it is far from free, and this source could emit a rather large number, particularly if cgroups is enabled.

Also note that that `timestamp` is set after each collection, immediately before generating the metrics from the scraped data, so it is quite close (except for cgroups, as noted there).

Hmm, having said that, `rust-timeit` is saying less than 50ns per call on my EC2 instance, and similar on my MacBook. That's less than I recall timing it at last time. Even if there are a couple thousand metrics, that's still in the 100µs range, which is relatively small for something that only happens every 2s. I could go either way on this if you're okay with that.

Reply:
Ah, yeah, that's fair, I always forget that there is some negligible overhead of fetching the time. It seems like you preserved the current behavior by setting it once per part of the collection so I think this is good to go in as-is.

Possibly relevant: https://github.com/vectordotdev/vector/issues/12451

Reply:
That's related, though I would expect the `stat` system calls to be at least a couple of orders of magnitude more costly.